README.md
# stroom-spark-datasource

[![Build Status](https://travis-ci.org/gchq/stroom-spark-datasource.svg?branch=master)](https://travis-ci.org/gchq/stroom-spark-datasource)

## Overview

This project provides an Apache Spark DataSource for Stroom.

Using this library it is possible to search Stroom indexes from Apache Spark, both via the DSL and via Spark SQL.

This DataSource is currently built for v2.4.8, and has been tested with v2.4.3, v2.4.7 and v2.4.8.
It is also likely to be compatible with other versions of Apache Spark v2.4.x

## Building

The project is built with gradle.

Running command `gradle build` will compile the software and build the library jar file.

However, in order to simplify deployment, it is possible to run the command `gradle build fatJar` in order to create a
single jar file that contains the library itself plus all its dependencies.     

Key facts:
* Output Directory: `build/libs`
* Library: `stroom-spark-datasource-VERSION.jar`
* Fat Jar: `stroom-spark-datasource-VERSION-all.jar`

## Getting Started

The simple demonstrations of this capability that are contained within this repository all rely on there being
a Stroom running on `localhost`, into which the standard Stroom content and sample data has been loaded by running the
command `gradle setupSampleData` from the directory containing your git clone of the `stroom` git repository. 

These demonstrations are designed to show the datasource operating in its simplest configuration, for clarity.
For this reason, security has been disabled within Stroom, by setting the property `authenticationRequired : false` within
`local.yml`.
A real deployment would require a user account to be created within Stroom and for it to be assigned an authentication
token, that is then passed in when instantiating the Datasource.

Install Apache Spark on `localhost`.  Spark currently requires Java 8, therefore this must be available and `$JAVA_HOME`
set to point at the Java 8 JDK.

On such a single spark node, the python CLI can be started from a Java 8 shell using 
the command 
```
pyspark --jars `pwd`/build/libs/stroom-spark-datasource-VERSION-all.jar
``` 
 **- remember to replace VERSION with the version of the library that you are actually using!**

Then from within pyspark shell, try

```python
from pyspark.sql.types import *

basicSchema = StructType([StructField("streamId", StringType(), True, \
    metadata={"get": "StreamId"}),\
    StructField("eventId", StringType(), True,\
    metadata={"get": "EventId"})])

basicDf = spark.read.format('stroom.spark.datasource.StroomDataSource').\
    load(token='eyJhbGciOiJSUzI1NiIsImtpZCI6IjhhM2I1OGNhLTk2ZTctNGFhNC05ZjA3LTQ0MDBhYWVkMTQ3MSJ9.eyJleHAiOjE2NDEzMTYyODMsInN1YiI6ImFkbWluIiwiaXNzIjoic3Ryb29tIiwiYXVkIjoiTlhDbXJyTGpQR2VBMVN4NWNEZkF6OUV2ODdXaTNucHRUbzZSdzVmTC5jbGllbnQtaWQuYXBwcy5zdHJvb20taWRwIn0.mU7gfhvbNFVzQ5RX86zcnvS6XZeNpYDZRJ5Zx2rPOuZsha15kJjMKw0PEH2T1Ucs2JEuhrj5P0Vu7pi_i7CtWSLNsmjZYrYL-udugijhEN9dxjuze5lnuOxb1LENrNoPaRvHpBiFyO3yIYO4Y9UcRdcAj-9Yt8zgqLP208KOU36n3dl0ErKXAKFyUbuu--1sDCnpSq4fxCMWfUhgENIkhUm6EnI8s510QdDxIeK95nHcM0H_I7cV0YXfpL4HgnsCCPJUJkYT1x71FQGuACi9Ba0JKdVF8PwdH4k4ryoK_b3ObFSueGpq2fBcFy385r6lDLUu91MODoY7yh9gagK-gw', \
    timestampField='Event Time', \
    host='localhost:8080', \
    protocol='http',uri='api/stroom-index/v2', \
    index='3e72379d-403f-4757-88d5-bfae4d724b55', \
    pipeline='2ac7b541-293d-4d9e-bb8d-f081dd631c7c', \
    schema=basicSchema)

basicDf.groupBy(basicDf['streamId']).count().sort(['count'], ascending=False).show()
```

When using a JSON Search extraction pipeline within Stroom, it is possible to work with the JSON directly in Spark.
```python
from pyspark.sql.functions import from_json, col

json_schema = spark.read.json(basicDf.rdd.map(lambda row: row.json)).schema
json_schema

structuredDf = basicDf.withColumn('evt', from_json(col('json'), json_schema))

wideDf=structuredDf.withColumn ('timestamp', col('evt.EventTime.TimeCreated')).\
    withColumn ('user', col('evt.EventSource.User.Id')).\
    withColumn('operation', col('evt.EventDetail.TypeId'))

wideDf.show()

wideDf.filter((wideDf['User'] == 'user1') | (wideDf['User'] == 'user2') | 
    (wideDf['User'] == 'user3')).groupBy(wideDf['Operation']).count().show()
```

If you are running a version of Stroom that supports `XPathOutputFilter`, XPaths can be used to access data directly.

```python
from pyspark.sql.types import *

xpathSchema = StructType([StructField("user", StringType(), True, \
    metadata={"get": "EventSource/User/Id"}),\
    StructField("operation", StringType(), True,\
    metadata={"get": "EventDetail/TypeId"})])

df = spark.read.format('stroom.spark.datasource.StroomDataSource').\
    load(token='eyJhbGciOiJSUzI1NiIsImtpZCI6IjhhM2I1OGNhLTk2ZTctNGFhNC05ZjA3LTQ0MDBhYWVkMTQ3MSJ9.eyJleHAiOjE2NDEzMTYyODMsInN1YiI6ImFkbWluIiwiaXNzIjoic3Ryb29tIiwiYXVkIjoiTlhDbXJyTGpQR2VBMVN4NWNEZkF6OUV2ODdXaTNucHRUbzZSdzVmTC5jbGllbnQtaWQuYXBwcy5zdHJvb20taWRwIn0.mU7gfhvbNFVzQ5RX86zcnvS6XZeNpYDZRJ5Zx2rPOuZsha15kJjMKw0PEH2T1Ucs2JEuhrj5P0Vu7pi_i7CtWSLNsmjZYrYL-udugijhEN9dxjuze5lnuOxb1LENrNoPaRvHpBiFyO3yIYO4Y9UcRdcAj-9Yt8zgqLP208KOU36n3dl0ErKXAKFyUbuu--1sDCnpSq4fxCMWfUhgENIkhUm6EnI8s510QdDxIeK95nHcM0H_I7cV0YXfpL4HgnsCCPJUJkYT1x71FQGuACi9Ba0JKdVF8PwdH4k4ryoK_b3ObFSueGpq2fBcFy385r6lDLUu91MODoY7yh9gagK-gw', \
    host='localhost:8080', \
    protocol='http', \
    uri='api/stroom-index/v2', \
    index='57a35b9a-083c-4a93-a813-fc3ddfe1ff44', \
    pipeline='26ed1000-255e-4182-b69b-00266be891ee', \
    schema=xpathSchema)

df.filter((df['user'] == 'user1') | (df['user'] == 'user2') | (df['user'] == 'user3')).\
    groupBy(df['operation']).count().show()

```




CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased

### Added

### Changed

### Removed

## [v2.0-alpha.1] - 20121-05-26

Replaced deprecated `stroom-query` client library with one based on stubs created using
 `openapi-generator`  / `swagger-codegen`
 

## [v1.0-alpha.4] - 2019-10-30
Equivalent of [v1.0-alpha.4_stroom-v6.x] but designed to be compatible with Stroom v7


## [v1.0-alpha.4_stroom-v6.x] - 2019-10-23

### Changed

Will never wait more than 60s between search intervals, to ensure result set kept alive

## [v1.0-alpha.3_stroom-v6.x] - 2019-10-23

### Added
Optional parameters for use with datasource:
* traceLevel: integer 0 (off/default) to 5 (max)
* maxResults: integer (default = 100000)
* pageSize: integer (default = 10000)
* numPartitions: integer (default = 3)

## [v1.0-alpha.2_stroom-v6.x] - 2019-10-21

* Initial alpha release, compatible with Stroom v6.x

## [v1.0-alpha.1] - 2019-10-14

* Initial alpha release

[Unreleased]: https://github.com/gchq/stroom-spark-datasource/compare/v1.0-alpha.3_stroom-v6.x...HEAD
[v1.0-alpha.3_stroom6.x]: https://github.com/gchq/stroom-spark-datasource/compare/v1.0-alpha.3_stroom-v6.x...v1.0-alpha.4_stroom-v6.x
[v1.0-alpha.3_stroom6.x]: https://github.com/gchq/stroom-spark-datasource/compare/v1.0-alpha.2_stroom-v6.x...v1.0-alpha.3_stroom-v6.x
[v1.0-alpha.2_stroom6.x]: https://github.com/gchq/stroom-spark-datasource/compare/v1.0-alpha.1...v1.0-alpha.2_stroom-v6.x
[v1.0-alpha.1]: https://github.com/gchq/stroom-spark-datasource/compare/v1.0-alpha.1...v1.0-alpha.1

src/main/java/stroom/spark/datasource/StroomDataSource.java
/*
 * Copyright 2019 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package stroom.spark.datasource;


import org.apache.spark.sql.sources.v2.DataSourceOptions;
import org.apache.spark.sql.sources.v2.DataSourceV2;
import org.apache.spark.sql.sources.v2.ReadSupport;
import org.apache.spark.sql.sources.v2.reader.DataSourceReader;
import org.apache.spark.sql.types.*;


public class StroomDataSource implements DataSourceV2, ReadSupport {
    public static final String INDEX_DOCREF_TYPE_ID = "Index";
    public static final String EXTRACTION_PIPELINE_DOCREF_TYPEID = "Pipeline";
    public static final String EXTRACTION_PIPELINE_NAME = "Extraction Pipeline For Spark Datasource";
    public static final String INDEX_NAME = "Index for Spark Datasource";

    public static final String FIELD_CONTENT_METADATA_KEY = "get";
    public static final String INDEXED_FIELD_METADATA_KEY = "indexField";
    public static final String AUTH_TOKEN_KEY = "token";
    public static final String HOST_KEY = "host";
    public static final String URL_PATH_KEY = "uri";
    public static final String DESTROY_PATH_KEY = "destroy_path";
    public static final String DATASOURCE_PATH_KEY = "datasource_path";
    public static final String SEARCH_PATH_KEY = "search_path";
    public static final String PROTOCOL_KEY = "protocol";
    public static final String EXTRACTION_PIPELINE_UUID_KEY = "pipeline";
    public static final String INDEX_UUID_KEY = "index";
    public static final String EVENT_TIME_FIELD_NAME_KEY = "timestampField";
    public static final String JSON_FIELD_NAME_KEY = "jsonField";
    public static final String TRACE_LEVEL_KEY = "traceLevel";
    public static final String MAX_RESULTS_KEY = "maxResults";
    public static final String PAGE_SIZE_KEY = "pageSize";
    public static final String NUM_PARTITIONS_KEY = "numPartitions";

    private String url = "indexService/v2";
    private String searchPath = "search";
    private String destroyPath = "destroy";
    private String datasourcePath = "dataSource";
    private String host = "localhost";
    private String protocol = "https";
    private String token = null;
    private String extractionPipelineUUID = null;
    private String indexUUID = null;
    private String eventTimeFieldName = "EventTime";
    private String jsonFieldName = "Json";
    private int traceLevel = 0;
    private String traceLevelStr = "" + traceLevel;
    private int maxResults = 100000;
    private String maxResultsStr = "" + maxResults;
    private int pageSize = 10000;
    private String pageSizeStr = "" + pageSize;
    private int numPartitions = 3;
    private String numPartitionsStr = "" + numPartitions;

    public DataSourceReader createReader(DataSourceOptions dataSourceOptions)
    {
        return createReader(null, dataSourceOptions);
    }

    /**
     *
     * @param schema - User provided schema.  Every field should have metadata with a defined key of xpath
     * @param dataSourceOptions Configuration passed by Spark during construction, contains user defined key/val map
     * @return newly created Reader
     */
    public DataSourceReader createReader(StructType schema, DataSourceOptions dataSourceOptions){
        token = dataSourceOptions.get(AUTH_TOKEN_KEY).orElseThrow(() ->
                new IllegalArgumentException("Authentication token must be provided via DataSourceOptions using key='"+
                        AUTH_TOKEN_KEY + "'"));
        extractionPipelineUUID = dataSourceOptions.get(EXTRACTION_PIPELINE_UUID_KEY).orElseThrow(() ->
                new IllegalArgumentException("Extraction Pipeline UUID must be provided via DataSourceOptions using key='"+
                        EXTRACTION_PIPELINE_UUID_KEY + "'"));
        indexUUID = dataSourceOptions.get(INDEX_UUID_KEY).orElseThrow(() ->
                new IllegalArgumentException("Index UUID of index to search must be provided via DataSourceOptions using key='"+
                        INDEX_UUID_KEY + "'"));


        host = dataSourceOptions.get(HOST_KEY).orElse(host);
        url = dataSourceOptions.get(URL_PATH_KEY).orElse(url);
        destroyPath = dataSourceOptions.get(DESTROY_PATH_KEY).orElse(destroyPath);
        searchPath = dataSourceOptions.get(SEARCH_PATH_KEY).orElse(searchPath);
        datasourcePath = dataSourceOptions.get(DATASOURCE_PATH_KEY).orElse(datasourcePath);
        protocol = dataSourceOptions.get(PROTOCOL_KEY).orElse(protocol);
        eventTimeFieldName = dataSourceOptions.get(EVENT_TIME_FIELD_NAME_KEY).orElse(eventTimeFieldName);
        jsonFieldName = dataSourceOptions.get(JSON_FIELD_NAME_KEY).orElse(jsonFieldName);

        traceLevelStr = dataSourceOptions.get(TRACE_LEVEL_KEY).orElse(traceLevelStr);
        try{
            traceLevel = Integer.parseInt(traceLevelStr);
        }catch (NumberFormatException ex){
            System.err.println("ERROR: Parameter " + TRACE_LEVEL_KEY + " must be assigned to a number (expressed as a string); using default of " + traceLevel);
        }

        maxResultsStr = dataSourceOptions.get(MAX_RESULTS_KEY).orElse(maxResultsStr);
        try{
            maxResults = Integer.parseInt(maxResultsStr);
        }catch (NumberFormatException ex){
            System.err.println("ERROR: Parameter " + MAX_RESULTS_KEY + " must be assigned to a number (expressed as a string); using default of " + maxResults);
        }

        pageSizeStr = dataSourceOptions.get(PAGE_SIZE_KEY).orElse(pageSizeStr);
        try{
            pageSize = Integer.parseInt(pageSizeStr);
        }catch (NumberFormatException ex){
            System.err.println("ERROR: Parameter " + PAGE_SIZE_KEY + " must be assigned to a number (expressed as a string); using default of " + pageSize);
        }

        numPartitionsStr = dataSourceOptions.get(NUM_PARTITIONS_KEY).orElse(numPartitionsStr);
        try{
            numPartitions = Integer.parseInt(numPartitionsStr);
        }catch (NumberFormatException ex){
            System.err.println("ERROR: Parameter " + NUM_PARTITIONS_KEY + " must be assigned to a number (expressed as a string); using default of " + numPartitions);
        }

        return new StroomDataSourceReader(schema, protocol, host, url,
                searchPath, destroyPath, datasourcePath,
                token,
                indexUUID, extractionPipelineUUID,
                eventTimeFieldName,jsonFieldName,
                numPartitions, pageSize, maxResults,
                traceLevel);
    }
}
